src/dstack/_internal/server/services/volumes.py

```diff
@@ -0,0 +1,259 @@
+    BackendNotAvailable,
+    ResourceExistsError,
+    ServerClientError,
+)
+from dstack._internal.core.models.volumes import (
+    Volume,
+    VolumeAttachmentData,
+    VolumeConfiguration,
+    VolumeProvisioningData,
+    VolumeStatus,
+)
+from dstack._internal.core.services import validate_dstack_resource_name
+from dstack._internal.server.models import ProjectModel, VolumeModel
+from dstack._internal.server.services import backends as backends_services
+from dstack._internal.server.utils.common import run_async, wait_to_lock_many
+from dstack._internal.utils import common, random_names
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+PROCESSING_VOLUMES_LOCK = asyncio.Lock()
+PROCESSING_VOLUMES_IDS = set()
+
+
+async def list_project_volumes(
+    session: AsyncSession,
+    project: ProjectModel,
+    names: Optional[List[str]] = None,
+) -> List[Volume]:
+    volume_models = await list_project_volume_models(session=session, project=project, names=names)
+    return [volume_model_to_volume(v) for v in volume_models]
+
+
+async def list_project_volume_models(
+    session: AsyncSession,
+    project: ProjectModel,
+    names: Optional[List[str]] = None,
+    include_deleted: bool = False,
+) -> List[VolumeModel]:
+    filters = [
+        VolumeModel.project_id == project.id,
+    ]
+    if names is not None:
+        filters.append(VolumeModel.name.in_(names))
+    if not include_deleted:
+        filters.append(VolumeModel.deleted == False)
+    res = await session.execute(select(VolumeModel).where(*filters))
+    return list(res.scalars().all())
+
+
+async def get_volume_by_name(
+    session: AsyncSession, project: ProjectModel, name: str
+) -> Optional[Volume]:
+    volume_model = await get_project_volume_model_by_name(
+        session=session, project=project, name=name
+    )
+    if volume_model is None:
+        return None
+    return volume_model_to_volume(volume_model)
+
+
+async def get_project_volume_model_by_name(
+    session: AsyncSession,
+    project: ProjectModel,
+    name: str,
+    include_deleted: bool = False,
+) -> Optional[VolumeModel]:
+    filters = [
+        VolumeModel.name == name,
+        VolumeModel.project_id == project.id,
+    ]
+    if not include_deleted:
+        filters.append(VolumeModel.deleted == False)
+    res = await session.execute(select(VolumeModel).where(*filters))
+    return res.scalar_one_or_none()
+
+
+async def create_volume(
+    session: AsyncSession,
+    project: ProjectModel,
+    configuration: VolumeConfiguration,
+) -> Volume:
+    _validate_volume_configuration(configuration)
+
+    if configuration.name is not None:
+        volume_model = await get_project_volume_model_by_name(
+            session=session,
+            project=project,
+            name=configuration.name,
+        )
+        if volume_model is not None:
+            raise ResourceExistsError()
+
+    if configuration.name is None:
```

Comment:
(nit) Two identical `if`s
```suggestion
```

Reply:
They are not identical: `is None` vs `is not None`. Maybe replace with else for readability...